asdf/schema.py

```diff
@@ -168,45 +168,52 @@ def validate_type(validator, types, instance, schema):
 })
 
 
-def validate_fill_default(validator, properties, instance, schema):
-    if not validator.is_type(instance, 'object'):
-        return
+_PROPERTIES_VALIDATOR = mvalidators.Draft4Validator.VALIDATORS['properties']
 
-    for property, subschema in properties.items():
-        if "default" in subschema:
-            instance.setdefault(property, subschema["default"])
 
-    for err in mvalidators.Draft4Validator.VALIDATORS['properties'](
-        validator, properties, instance, schema):
-        yield err
+def validate_fill_default(validator, properties, instance, schema):
+    if validator.is_type(instance, 'object'):
+        for property, subschema in properties.items():
+            if "default" in subschema:
+                instance.setdefault(property, subschema["default"])
+
+    yield from _PROPERTIES_VALIDATOR(validator, properties, instance, schema)
 
 
+# Used when filling defaults in a separate validation pass:
 FILL_DEFAULTS = util.HashableDict()
-for key in ('allOf', 'anyOf', 'oneOf', 'items'):
+for key in ('allOf', 'anyOf', 'oneOf', 'items', '$ref'):
```

Comment:
Adding `$ref` here is a bug fix on the existing defaults behavior, which until now has ignored defaults from referenced schemas.  I noticed the issue when comparing to the result of the combined defaults/validation pass, which gave a different answer that turned out to be correct.